### ISSUE TITLE
topojson.mesh filter doesn’t support geometry with multiple neighbors

### ISSUE THREAD

**Issue by evanl5** [topojson/topojson-client#9]
If there are more than two geometries in a collection that share the same arcs, the mesh filter function will ignore some neighboring geometries.

For instance, given the following topology, the filter function will never be passed geometry with id 2.

```javascript
var topology = {
  "type": "Topology",
  "objects": {
    "collection": {
      "type": "GeometryCollection",
      "geometries": [
        {"id": 1, "type": "LineString", "arcs": [0,1]},
        {"id": 2, "type": "LineString", "arcs": [0,1]},
        {"id": 3, "type": "LineString", "arcs": [0,1]}
      ]
    }
  },
  "arcs": [
    [[1, 0], [2, 0]],
    [[0, 0], [1, 0]]
  ]
};

topojson.mesh(topology, topology.objects.collection, function (geometry, neighbor) {
  return geometry.id === 2 || neighbor.id === 2;
});
```

For a more practical use case, in the screenshot below the bold line is created with a mesh filter intend to outline Uganda. However, since Uganda shares borders with both Kenya and Kenya’s states, the filter fails and the border is partially missing.

![screen shot 2017-03-17 at 8 45 33 am](https://cloud.githubusercontent.com/assets/3799759/24063250/2b3e2a0c-0b1c-11e7-8c0b-8899c0f5badf.png)

The issue appears to be in the extractArcs function (https://github.com/topojson/topojson-client/blob/master/src/mesh.js#L50) where only the first and last neighbors are sent to the filter function.

Here are some possible solutions that I could think of:
- pass all geometries as a third argument to the filter function. 
```javascript 
function(geoms) { if (filter(geoms[0].g, geoms[geoms.length - 1].g, geoms.map(function(geom) { return geom.g; })) arcs.push(geoms[0].i); });
```
- only pass all geometries to the filter function. 
```javascript 
function(geoms) { if (filter(geoms.map(function(geom) { return geom.g; }))) arcs.push(geoms[0].i); });
```
- pass each geometry as an individual argument to the filter function
```javascript 
function(geoms) { if (filter.apply(this, geoms.map(function(geom) { return geom.g; }))) arcs.push(geoms[0].i); });
```

- allow the way the filter function is applied to be changed as an optional argument to the mesh function

```javascript
topojson.mesh(topology, object, filter, [apply = defaultApply])

function defaultApply(geometries, filterFn) {
  return geometries.reduce(function (accum, geom, i, geoms) {
    if (accum === false) return false;
    return filterFn(geoms[0], geom);
  }, true);
}

function (geoms) { 
  var apply = apply || defaultApply;
  if(apply(geoms.map(function(geom) { return geom.g; }), filter) arcs.push(geoms[0].i); 
});
```

**Comment by mbostock** [comment]
This is the expected behavior but there is an unstated assumption that the topology used to derive the mesh is valid, part of which is a requirement that the geometry used to derive the mesh is not self-overlapping. So in your example, you’ll need to decide whether to derive the mesh from Kenya or Kenya’s states, but you can’t include both in the source geometry; I recommend grouping your geometry by administrative level, as is done for example in [us-atlas](https://github.com/topojson/us-atlas), which provides separate layers for the nation, states and counties.

**Comment by evanl5** [comment]
My geometry actually is grouped by administrative level. I have four administrative levels and three desired mesh layers, creating them as separate layers would require 12 layers. In an effort to simplify the process of creating the mesh layers I decided to pass all visible administrative levels to the mesh function at once as nested GeometryCollections. This is how I discovered the problem.

Is there any chance you would consider a feature request to add this functionality?
